boot/setefibootvars_linux.go

```diff
@@ -0,0 +1,210 @@
+// finds the first unused boot variable number and uses it. Writes the load
+// option to that variable, and returns the number of the variable that was
+// used.
+func setEfiBootOptionVariable(loadOptionData []byte) (uint16, error) {
+	bootNum, err := findMatchingBootOption(loadOptionData)
+	if err == nil {
+		return bootNum, nil
+	} else if err != ErrNoMatchingVariable {
+		return 0, err
+	}
+	varName := fmt.Sprintf("Boot%04X", bootNum)
+	err = efiWriteVariable(varName, efi.GlobalVariable, defaultVarAttrs, loadOptionData)
+	return bootNum, err
+}
+
+// setEfiBootOrderVariable reads the current BootOrder variable,
+// inserts the given newBootNum at the beginning of the number list
+// (and removes it from later in the list if it occurs) and writes the
+// list as the new BootOrder variable.
+func setEfiBootOrderVariable(newBootNum uint16) error {
+	origData, attrs, err := efiReadVariable("BootOrder", efi.GlobalVariable)
+	if err == efi.ErrVarNotExist {
+		attrs = defaultVarAttrs
+		origData = make([]byte, 0)
+	} else if err != nil {
+		return err
+	}
+	if len(origData)%2 != 0 {
+		return ErrInvalidBootOrder
+	}
+	var optionOffset = -1
+	for i := 0; i < len(origData); i += 2 {
+		bootNum := binary.LittleEndian.Uint16(origData[i : i+2])
+		if newBootNum == bootNum {
+			optionOffset = i
+			break
+		}
+	}
+	var newData []byte
+	if optionOffset == 0 {
+		// newBootNum already at start, no need to re-write variable
+		return nil
+	} else if optionOffset == -1 {
+		// newBootNum not in original boot order
+		newData = make([]byte, len(origData)+2)
+		binary.LittleEndian.PutUint16(newData, newBootNum)
+		copy(newData[2:], origData)
+	} else {
+		newData = make([]byte, len(origData))
+		binary.LittleEndian.PutUint16(newData, newBootNum)
+		copy(newData[2:], origData[:optionOffset])
+		copy(newData[optionOffset+2:], origData[optionOffset+2:])
+	}
+	return efiWriteVariable("BootOrder", efi.GlobalVariable, attrs, newData)
+}
+
+// SetEfiBootVariables sets the Boot#### and BootOrder variables for the given
+// load option information.
+//
+// Constructs an EFI load option with the given description, the device path
+// corresponding to the given asset path, and the given optional data. Writes
+// the EFI boot variable Boot#### to contain the resulting load option. Then,
+// sets the BootOrder variable so that the #### number from the chosen Boot####
+// is first, removing it from elsewhere in the BootOrder if it occurs.
+func setEfiBootVariablesImpl(description string, assetPath string, optionalData []byte) error {
+	loadOptionData, err := constructLoadOption(description, assetPath, optionalData)
+	if err != nil {
+		return err
+	}
+	bootNum, err := setEfiBootOptionVariable(loadOptionData)
+	if err != nil {
+		return err
+	}
+	return setEfiBootOrderVariable(bootNum)
+}
+
+var SetEfiBootVariables = setEfiBootVariablesImpl
+
+func MockSetEfiBootVariables(f func(description string, assetPath string, optionalData []byte) error) func() {
```

Comment:
nitpick: Is this the only way to allow this to be mocked from other test suites? Perhaps a comment saying not to use this in non-test code, but probably unnecessary.

Reply:
I think so. It is used `overlord/devicestate`. So that means `export_test.go` would not work. I think we have around 150 such function around the code. I would not think to explain what it is for from the name. @bboozzoo or @pedronis probably have an opinion about this.